### PR TITLE
Deflake `operator/garden/garden` integration test

### LINKED ISSUE #10576: [Flaky Test] Garden controller tests [It] should successfully reconcile and delete a Garden
**How to categorize this issue?**

/area testing
/kind flake

**Which test(s)/suite(s) are flaking**:
- Garden controller tests [It] should successfully reconcile and delete a Garden

**CI link**:
- https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/10564/pull-gardener-integration/1839195831954051072
- https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/10472/pull-gardener-integration/1838993964762402816

**Reason for failure**:
```
 [FAILED] Timed out after 10.001s.
  Expected success, but got an error:
      <*errors.StatusError | 0xc00444d180>: 
      servicemonitors.monitoring.coreos.com "garden-gardener-operator" not found
      {
          ErrStatus: {
              TypeMeta: {Kind: "", APIVersion: ""},
              ListMeta: {
                  SelfLink: "",
                  ResourceVersion: "",
                  Continue: "",
                  RemainingItemCount: nil,
              },
              Status: "Failure",
              Message: "servicemonitors.monitoring.coreos.com \"garden-gardener-operator\" not found",
              Reason: "NotFound",
              Details: {
                  Name: "garden-gardener-operator",
                  Group: "monitoring.coreos.com",
                  Kind: "servicemonitors",
                  UID: "",
                  Causes: nil,
                  RetryAfterSeconds: 0,
              },
              Code: 404,
          },
      }
  In [It] at: /home/prow/go/src/github.com/gardener/gardener/test/integration/operator/garden/garden/garden_test.go:429 @ 09/26/24 06:56:54.405
```

**Anything else we need to know**:
N/A

### PR DESCRIPTION
**Pull request by tobschli**
**How to categorize this PR?**

/area dev-productivity
/kind flake

**What this PR does / why we need it**:

This PR aims to fix the flaky test mentioned in #10576.

This flake seems to be related to #10464. Within the operator reconciliation flow, the `prometheus-operator` CRDs are deployed without waiting for them to be ready. As a result, when the kube-api-server is under heavy load, these CRDs may not be fully processed in time to deploy the `ServiceMonitor` through the same reconciliation flow.

This results in the api-server not knowing the `ServiceMonitor` resource, resulting in a reconciliation failure. The flow does not restart fast enough with the given timeout of `10s`, resulting in the timeout.

The changes modify the `prometheus-operator` and all other `crdDeployer` structs to `DeployWaiter` and the reconciliation flows to wait for the CRDs to be ready.

Where applicable, a generic `CRDDeployer` is being used. For the remaining `CRDDeployers` generic methods were used as much as possible.

In addition, the timeout of the `operator/garden/garden` integration test has been increased to `15s`, since under load the CRDs and subsequently the `ServiceMonitor` need more time to deploy successfully.

For verification, I stress tested the affected integration tests (`operator/garden/garden` and `gardenlet/seed/seed`) in a remote-local setup with `6` parallel processes and successfully passed `>60` test runs without errors.

**Which issue(s) this PR fixes**:
Fixes #10576

**Special notes for your reviewer**:

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/component/autoscaling/hvpa/crd.go [hunk 1]
```diff
@@ -7,23 +7,31 @@ package hvpa
 import (
 	"context"
 	_ "embed"
+	"time"
 
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
```
Comment (rfranzke) [inline pkg/component/autoscaling/hvpa/crd.go]:
```suggestion
	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
```

Maybe you can also add something like https://github.com/gardener/gardener/blob/ddde6b6121857c3a68cbcb88a4638d59930c79e2/.golangci.yaml.in#L69-L70 to enable linting

File: pkg/component/autoscaling/vpa/crd.go [hunk 2]
```diff
@@ -71,3 +79,46 @@ func (v *vpaCRD) Destroy(ctx context.Context) error {
...
+)
+
+// Wait signals whether a CRD is ready or needs more time to be deployed.
+func (v *vpaCRD) Wait(ctx context.Context) error {
+	timeoutCtx, cancel := context.WithTimeout(ctx, TimeoutWaitForCRD)
+	defer cancel()
+
+	return retry.Until(timeoutCtx, IntervalWaitForCRD, func(ctx context.Context) (done bool, err error) {
+		for _, resource := range crdResources {
+			r := resource
```
Comment (rfranzke) [inline pkg/component/autoscaling/vpa/crd.go]:
```suggestion
```

This line is not needed

File: pkg/component/autoscaling/vpa/crd.go [hunk 3]
```diff
@@ -71,3 +79,46 @@ func (v *vpaCRD) Destroy(ctx context.Context) error {
...
+			if err := v.client.Get(ctx, client.ObjectKeyFromObject(obj), crd); client.IgnoreNotFound(err) != nil {
+				return retry.SevereError(err)
+			}
+
+			if err := health.CheckCustomResourceDefinition(crd); err != nil {
+				return retry.MinorError(err)
+			}
+		}
+		return retry.Ok()
+	})
```
Comment (rfranzke) [inline pkg/component/autoscaling/vpa/crd.go]: Can you put this `func` somewhere to the utils to make it reusable? Now, it's copied and duplicated too often.

File: pkg/component/autoscaling/vpa/crd.go [hunk 4]
```diff
@@ -71,3 +79,46 @@ func (v *vpaCRD) Destroy(ctx context.Context) error {
...
+var (
+	// IntervalWaitForCRD is the interval used while waiting for the CRDs to become healthy
+	// or deleted.
+	IntervalWaitForCRD = 1 * time.Second
+	// TimeoutWaitForCRD is the timeout used while waiting for the CRDs to become healthy
+	// or deleted.
+	TimeoutWaitForCRD = 15 * time.Second
+	// Until is an alias for retry.Until. Exposed for tests.
+	Until = retry.Until
+)
```
Comment (rfranzke) [inline pkg/component/autoscaling/vpa/crd.go]: What's the benefit of these constants if there are no unit tests where they could get overridden?

**Review by LucaBernstein — COMMENTED** [review]
/lgtm

File: pkg/utils/kubernetes/crd.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/kubernetes/health"
+	"github.com/gardener/gardener/pkg/utils/retry"
+)
+
+// WaitUntilCRDManifestsReady takes manifests as strings and waits for them to get ready with a timeout of 15 seconds
+func WaitUntilCRDManifestsReady(ctx context.Context, c client.Client, manifests []string) error {
```
Comment (timuthy) [inline pkg/utils/kubernetes/crd.go]: When we spoke, didn't you plan on working with CRD names here instead, so unmarshalling is not required again?

Also, unit tests are missing for this function.

File: pkg/utils/kubernetes/crd.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
+// SPDX-FileCopyrightText: SAP SE or an SAP affiliate company and Gardener contributors
```
Comment (timuthy) [inline pkg/utils/kubernetes/crd.go]:
```suggestion
// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
```

Not sure if both is possible, though. (cc @rfranzke)

**Review by timuthy — CHANGES_REQUESTED** [review]

File: test/integration/operator/garden/garden/garden_suite_test.go [hunk 7]
```diff
@@ -47,7 +47,7 @@ var (
 var _ = BeforeSuite(func() {
 	// a lot of CPU-intensive stuff is happening in this test, so to
 	// prevent flakes we have to increase the timeout here manually
-	SetDefaultEventuallyTimeout(10 * time.Second)
+	SetDefaultEventuallyTimeout(15 * time.Second)
```
Comment (rfranzke) [inline test/integration/operator/garden/garden/garden_suite_test.go:50]: Is this change really needed?

Comment (tobschli) [reply test/integration/operator/garden/garden/garden_suite_test.go:50]: Unfortunately, yes.
Otherwise the test will often time out under load.

File: pkg/utils/kubernetes/crd.go (hunk 6 above)
Comment (tobschli) [reply pkg/utils/kubernetes/crd.go]: In the repository, I see both represented. I used the Makefile target ` add-license-headers` to generate it.
Which one should I use?

Comment (rfranzke) [reply pkg/utils/kubernetes/crd.go]: If `make add-license-headers` added it this way, I'd say it's fine for now.

File: pkg/component/autoscaling/hvpa/crd.go [hunk 8]
```diff
@@ -8,28 +8,82 @@ import (
...
+
+func init() {
+	resourceObjectKeyMap = make(map[client.ObjectKey]string)
+	resources := []string{crdHvpas}
+
+	for _, resource := range resources {
+		objKey, err := kubernetesutils.GetObjectKeyFromManifest(resource)
+		if err != nil {
+			panic(err)
+		}
```
Comment (timuthy) [inline pkg/component/autoscaling/hvpa/crd.go]:
```suggestion
		utilruntime.Must(err)
```

Same for other occurrences.

File: pkg/component/autoscaling/hvpa/crd.go [hunk 9]
```diff
@@ -8,28 +8,82 @@ import (
...
 )
 
-//go:embed templates/crd-autoscaling.k8s.io_hvpas.yaml
-var crd string
+var (
+	//go:embed templates/crd-autoscaling.k8s.io_hvpas.yaml
+	crdHvpas string
+
+	// resourceObjectKeyMap maps the ObjectKey of the CRD to its corresponding manifest
+	resourceObjectKeyMap map[client.ObjectKey]string
```
Comment (timuthy) [inline pkg/component/autoscaling/hvpa/crd.go]: A slice should be fine here? I don't see a benefit using a map here, and the need to copy the crd content.

File: pkg/component/autoscaling/vpa/crd.go [hunk 10]
```diff
@@ -20,23 +23,36 @@ var (
 	verticalPodAutoscalerCheckpointCRD string
 
 	crdResources map[string]string
+	// resourceObjectKeys is a slice of the CRD object keys
+	resourceObjectKeys []client.ObjectKey
```
Comment (timuthy) [inline pkg/component/autoscaling/vpa/crd.go]:
```suggestion
	crdNames []string
```

No doc string required if the var name is clear :)

File: pkg/component/certmanagement/crd.go [hunk 11]
```diff
@@ -23,26 +25,41 @@ var (
 	//go:embed assets/crd-cert.gardener.cloud_issuers.yaml
 	crdIssuers string
 
-	resources []string
+	// resourceObjectKeyMap maps the ObjectKey of the CRD to its corresponding manifest
+	resourceObjectKeyMap map[client.ObjectKey]string
```
Comment (timuthy) [inline pkg/component/certmanagement/crd.go]:
```suggestion
	crdNameToResource map[string]string
```

File: pkg/component/etcd/etcd/crd.go [hunk 12]
```diff
@@ -20,18 +21,20 @@ import (
...
 	gardenerutils "github.com/gardener/gardener/pkg/utils/gardener"
+	kubernetesutils "github.com/gardener/gardener/pkg/utils/kubernetes"
 )
 
 var (
 	//go:embed crds/templates/crd-druid.gardener.cloud_etcds.yaml
-	// CRD holds the etcd custom resource definition template
-	CRD string
+	// CrdCloudEtcds holds the etcd custom resource definition template
+	CrdCloudEtcds string
```
Comment (timuthy) [inline pkg/component/etcd/etcd/crd.go]: Why is this changed and what is a CloudEtcd? 😃 

Please revert.

File: pkg/component/etcd/etcd/crd.go [hunk 13]
```diff
@@ -20,18 +21,20 @@ import (
...
 	//go:embed crds/templates/crd-druid.gardener.cloud_etcdcopybackupstasks.yaml
-	etcdCopyBackupsTaskCRD string
+	crdEtcdCopyBackupsTasks string
 
 	etcdCRDName                = "etcds.druid.gardener.cloud"
 	etcdCopyBackupsTaskCRDName = "etcdcopybackupstasks.druid.gardener.cloud"
-	crdResources               []string
+	// resourceObjectKeyMap maps the ObjectKey of the CRD to its corresponding manifest
+	resourceObjectKeyMap map[client.ObjectKey]string
```
Comment (timuthy) [inline pkg/component/etcd/etcd/crd.go]:
```suggestion
	crdNameToResource map[string]string
```

For remaining occurrences as well.

File: pkg/gardenlet/controller/seed/seed/components.go [hunk 14]
```diff
@@ -73,14 +73,14 @@ import (
...
-	etcdCRD       component.Deployer
-	istioCRD      component.Deployer
-	vpaCRD        component.Deployer
-	hvpaCRD       component.Deployer
-	fluentCRD     component.Deployer
-	prometheusCRD component.Deployer
+	machineCRD    component.DeployWaiter
+	extensionCRD  component.DeployWaiter
+	etcdCRD       component.DeployWaiter
+	istioCRD      component.DeployWaiter
```
Comment (timuthy) [inline pkg/gardenlet/controller/seed/seed/components.go:79]: Wait is not implemented yet for `istioCRD`.

File: pkg/gardenlet/controller/seed/seed/components.go [hunk 15]
```diff
@@ -73,14 +73,14 @@ import (
...
-	machineCRD    component.Deployer
-	extensionCRD  component.Deployer
-	etcdCRD       component.Deployer
-	istioCRD      component.Deployer
-	vpaCRD        component.Deployer
-	hvpaCRD       component.Deployer
-	fluentCRD     component.Deployer
-	prometheusCRD component.Deployer
+	machineCRD    component.DeployWaiter
+	extensionCRD  component.DeployWaiter
```
Comment (timuthy) [inline pkg/gardenlet/controller/seed/seed/components.go:77]: Wait is not implemented yet for `extensionCRD`.

File: pkg/utils/kubernetes/crd.go [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+	"github.com/gardener/gardener/pkg/utils/retry"
+)
+
+var (
+	// CRDWaitTimeout specifies the total time to wait for CRDs to become ready.
+	CRDWaitTimeout = 15 * time.Second
+)
+
+// WaitUntilCRDManifestsReady takes CRD ObjectKeys and waits for them to get ready with a timeout of 15 seconds
+func WaitUntilCRDManifestsReady(ctx context.Context, c client.Client, crdObjectKeys []client.ObjectKey) error {
```
Comment (timuthy) [inline pkg/utils/kubernetes/crd.go]:
```suggestion
func WaitUntilCRDManifestsReady(ctx context.Context, c client.Client, names []string) error {
```

File: pkg/utils/kubernetes/crd_test.go [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+				},
+			},
+		}
+
+		ValidManifest = `apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+    name: myresources.mygroup.example.com`
+		InValidManifest = `thisIsNotAValidManifest`
+	)
```
Comment (timuthy) [inline pkg/utils/kubernetes/crd_test.go]: Can you assign those values in `BeforeEach`, to ensure each test has an initial state?

File: pkg/utils/kubernetes/crd_test.go [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+			CRDWaitTimeout = 100 * time.Millisecond
+			fakeClient := fakeclient.NewClientBuilder().
+				WithScheme(apiextensionsscheme.Scheme).
+				WithObjects(UnreadyCrd).
+				Build()
+
+			err := WaitUntilCRDManifestsReady(ctx, fakeClient, []cli.ObjectKey{cli.ObjectKeyFromObject(UnreadyCrd)})
+
+			Expect(err).To(HaveOccurred())
+			Expect(err).To(MatchError(ContainSubstring("context deadline exceeded")))
```
Comment (timuthy) [inline pkg/utils/kubernetes/crd_test.go]: Maybe check the entire string to be more precise and provider clarity in the test?

**Review by timuthy — CHANGES_REQUESTED** [review]

File: pkg/component/autoscaling/hvpa/crd.go (hunk 9 above)
Comment (tobschli) [reply pkg/component/autoscaling/hvpa/crd.go]: In this example. actually a simple strong variable would suffice, but I thought aligning it with the other implementations prevents confusion.
Should I revert it back to its original form?

File: pkg/component/autoscaling/vpa/crd.go (hunk 10 above)
Comment (tobschli) [reply pkg/component/autoscaling/vpa/crd.go]: I would still prefer keeping it a `client.ObjectKey` slice to prevent unmarshalling the resources on every `Wait()`

File: pkg/component/certmanagement/crd.go (hunk 11 above)
Comment (tobschli) [reply pkg/component/certmanagement/crd.go]: The name suggestion sounds good, but I still would prefer the `client.ObjectKey` as key for the map.

File: pkg/component/etcd/etcd/crd.go (hunk 12 above)
Comment (tobschli) [reply pkg/component/etcd/etcd/crd.go]: I aligned the variable name to the other packages, aas the naming follows a pattern of the filename, in this case `cloud_etcds.yaml`, therefore I aligned it.
Can also revert it back, if needed.

File: pkg/utils/kubernetes/crd_test.go (hunk 18 above)
Comment (tobschli) [reply pkg/utils/kubernetes/crd_test.go]: The entire string is quite bulky, and the `context deadline exceeded` error is embedded in a `retry` error, which could be any kind of error.

File: pkg/component/certmanagement/crd.go [hunk 19]
```diff
@@ -24,70 +21,10 @@ var (
...
-
-// Wait signals whether a CRD is ready or needs more time to be deployed.
-func (c *crdDeployer) Wait(ctx context.Context) error {
-	return kubernetesutils.WaitUntilCRDManifestsReady(ctx, c.client, maps.Keys(resourceObjectKeyMap))
-}
-
-// WaitCleanup for destruction to finish and component to be fully removed. crdDeployer does not need to wait for cleanup.
-func (c *crdDeployer) WaitCleanup(_ context.Context) error {
-	return nil
+	resources := []string{crdRevocations, crdCertificates, crdIssuers}
```
Comment (rfranzke) [inline pkg/component/certmanagement/crd.go]: Eliminate this useless variable

File: pkg/utils/kubernetes/crd.go [hunk 20]
```diff
@@ -21,16 +24,80 @@ var (
...
+		crdNameToManifest: MakeCrdNameMap(manifests),
+	}
+}
+
+// Deploy deploys the CRDs
+func (c *CRDDeployer) Deploy(ctx context.Context) error {
+	var fns []flow.TaskFn
+
+	for _, resource := range c.crdNameToManifest {
+		r := resource
```
Comment (rfranzke) [inline pkg/utils/kubernetes/crd.go]: I don't think this line is necessary anymore

File: pkg/utils/kubernetes/crd.go [hunk 21]
```diff
@@ -21,16 +24,80 @@ var (
...
+		fns = append(fns, func(ctx context.Context) error {
+			return c.applier.ApplyManifest(ctx, kubernetes.NewManifestReader([]byte(r)), kubernetes.DefaultMergeFuncs)
+		})
+	}
+
+	return flow.Parallel(fns...)(ctx)
+}
+
+// Destroy destroys the CRDs
```
Comment (rfranzke) [inline pkg/utils/kubernetes/crd.go]:
```suggestion
// Destroy destroys the CRDs.
```

File: pkg/utils/kubernetes/crd.go [hunk 22]
```diff
@@ -21,16 +24,80 @@ var (
...
+// NewCRDDeployer returns a DeployWaiter that can deploy CRDs and wait for them to be ready
+func NewCRDDeployer(client client.Client, applier kubernetes.Applier, manifests []string) *CRDDeployer {
+	return &CRDDeployer{
+		client:            client,
+		applier:           applier,
+		crdNameToManifest: MakeCrdNameMap(manifests),
+	}
+}
+
+// Deploy deploys the CRDs
```
Comment (rfranzke) [inline pkg/utils/kubernetes/crd.go]:
```suggestion
// Deploy deploys the CRDs.
```

File: pkg/utils/kubernetes/crd.go [hunk 23]
```diff
@@ -21,16 +24,80 @@ var (
...
+	return WaitUntilCRDManifestsReady(ctx, c.client, maps.Keys(c.crdNameToManifest))
+}
+
+// WaitCleanup for destruction to finish and component to be fully removed. crdDeployer does not need to wait for cleanup.
+func (c *CRDDeployer) WaitCleanup(_ context.Context) error {
+	return nil
+}
+
+// MakeCrdNameMap returns a map that has the name of the resource as key, and the corresponding manifest as value
+func MakeCrdNameMap(manifests []string) map[string]string {
```
Comment (rfranzke) [inline pkg/utils/kubernetes/crd.go]:
```suggestion
// MakeCRDNameMap returns a map that has the name of the resource as key, and the corresponding manifest as value.
func MakeCRDNameMap(manifests []string) map[string]string {
```

File: pkg/utils/kubernetes/crd.go [hunk 24]
```diff
@@ -21,16 +24,80 @@ var (
...
+
+// Wait signals whether a CRD is ready or needs more time to be deployed.
+func (c *CRDDeployer) Wait(ctx context.Context) error {
+	return WaitUntilCRDManifestsReady(ctx, c.client, maps.Keys(c.crdNameToManifest))
+}
+
+// WaitCleanup for destruction to finish and component to be fully removed. crdDeployer does not need to wait for cleanup.
+func (c *CRDDeployer) WaitCleanup(_ context.Context) error {
+	return nil
+}
```
Comment (rfranzke) [inline pkg/utils/kubernetes/crd.go]: Why not? For the sake of completeness, shouldn't we implement this?

File: pkg/utils/kubernetes/crd.go [hunk 25]
```diff
@@ -21,16 +24,80 @@ var (
...
+
+// MakeCrdNameMap returns a map that has the name of the resource as key, and the corresponding manifest as value
+func MakeCrdNameMap(manifests []string) map[string]string {
+	crdNameToManifest := make(map[string]string)
+	for _, manifest := range manifests {
+		name, err := GetObjectNameFromManifest(manifest)
+		utilruntime.Must(err)
```
Comment (rfranzke) [inline pkg/utils/kubernetes/crd.go]: Don't panic

File: pkg/utils/kubernetes/crd_test.go [hunk 26]
```diff
@@ -32,15 +39,28 @@ var _ = Describe("CRD", func() {
 			},
 		}
 
+		ValidManifest   string
+		InvalidManifest string
```
Comment (rfranzke) [inline pkg/utils/kubernetes/crd_test.go]: Why are these variable exported?

File: test/integration/gardenlet/seed/seed/seed_suite_test.go [hunk 27]
```diff
@@ -90,6 +91,7 @@ var _ = BeforeSuite(func() {
 		kubernetes.AddSeedSchemeToScheme,
 		gardencorev1beta1.AddToScheme,
 		seedmanagementv1alpha1.AddToScheme,
+		apiextensionsv1.AddToScheme,
```
Comment (rfranzke) [inline test/integration/gardenlet/seed/seed/seed_suite_test.go]: Why is this suddenly needed?

File: test/integration/resourcemanager/crddeletionprotection/crddeletionprotection_test.go [hunk 28]
```diff
@@ -70,11 +72,15 @@ var _ = Describe("Extension CRDs Webhook Handler", func() {
 
 		By("Apply CRDs")
 		applier, err := kubernetes.NewApplierForConfig(restConfig)
+
+		s := runtime.NewScheme()
+		c := fake.NewClientBuilder().WithScheme(s).Build()
```
Comment (rfranzke) [inline test/integration/resourcemanager/crddeletionprotection/crddeletionprotection_test.go]: ... :)

File: pkg/utils/kubernetes/crd.go [hunk 29]
```diff
@@ -21,16 +24,80 @@ var (
...
 
+// CRDDeployer is a DeployWaiter that can deploy CRDs and wait for them to be ready
+type CRDDeployer struct {
+	client            client.Client
+	applier           kubernetes.Applier
+	crdNameToManifest map[string]string
+}
+
+// NewCRDDeployer returns a DeployWaiter that can deploy CRDs and wait for them to be ready
+func NewCRDDeployer(client client.Client, applier kubernetes.Applier, manifests []string) *CRDDeployer {
```
Comment (rfranzke) [inline pkg/utils/kubernetes/crd.go]: Unit tests missing

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/utils/kubernetes/crd.go (hunk 25 above)
Comment (tobschli) [reply pkg/utils/kubernetes/crd.go]: 😰😰😰😰😰

File: pkg/utils/kubernetes/crd.go [hunk 30]
```diff
@@ -0,0 +1,138 @@
...
+		}
+		err := WaitUntilResourceDeleted(timeoutCtx, c.client, crd, 1*time.Second)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// MakeCRDNameMap returns a map that has the name of the resource as key, and the corresponding manifest as value
```
Comment (LucaBernstein) [inline pkg/utils/kubernetes/crd.go]:
```suggestion
// MakeCRDNameMap returns a map that has the name of the resource as key, and the corresponding manifest as value.
```

Please also check the other Go Doc comments for missing sentence terminations.

**Review by LucaBernstein — COMMENTED** [review]
lgtm

File: pkg/utils/kubernetes/crd.go (hunk 30 above)
Comment (tobschli) [reply pkg/utils/kubernetes/crd.go]: Oh, thank you for pointing out my oversight :)

File: pkg/utils/kubernetes/crd.go [hunk 31]
```diff
@@ -0,0 +1,138 @@
...
+	defer cancel()
+
+	for resourceName, _ := range c.crdNameToManifest {
+		crd := &apiextensionsv1.CustomResourceDefinition{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: resourceName,
+			},
+		}
+		err := WaitUntilResourceDeleted(timeoutCtx, c.client, crd, 1*time.Second)
+		if err != nil {
```
Comment (rfranzke) [inline pkg/utils/kubernetes/crd.go]:
```suggestion
		if err := WaitUntilResourceDeleted(timeoutCtx, c.client, crd, 1*time.Second); err != nil {
```

File: pkg/utils/kubernetes/crd.go [hunk 32]
```diff
@@ -0,0 +1,138 @@
...
+			return err
+		}
+	}
```
Comment (rfranzke) [inline pkg/utils/kubernetes/crd.go]: I think it's better to use `flow.Parallel` and wait for all CRDs in parallel, and then return the error. This way, we wait for them one-by-one, which is not quite a nice UX in case they hang (it's better to get a full overview of all CRDs).

File: pkg/component/etcd/etcd/crd.go [hunk 33]
```diff
@@ -104,3 +110,13 @@ func (c *crd) Destroy(ctx context.Context) error {
...
 }
+
+// Wait signals whether a CRD is ready or needs more time to be deployed.
+func (c *crd) Wait(ctx context.Context) error {
+	return kubernetesutils.WaitUntilCRDManifestsReady(ctx, c.client, maps.Keys(crdNameToManifest))
+}
+
+// WaitCleanup for destruction to finish and component to be fully removed. crdDeployer does not need to wait for cleanup.
+func (c *crd) WaitCleanup(_ context.Context) error {
+	return nil
```
Comment (rfranzke) [inline pkg/component/etcd/etcd/crd.go]: Why not implemented? You could similarly extract the `WaitCleanup` function from the `CRDDeployer` and call it here.

File: pkg/component/extensions/crds/crds.go [hunk 34]
```diff
@@ -124,12 +134,19 @@ func (c *crd) Destroy(ctx context.Context) error {
...
+	if c.includeShootCRDs {
+		names = append(names, maps.Keys(shootCRDNamesToManifest)...)
+	}
+	return kubernetesutils.WaitUntilCRDManifestsReady(ctx, c.client, names)
 }
 
-// WaitCleanup does nothing
+// WaitCleanup does nothing.
 func (c *crd) WaitCleanup(_ context.Context) error {
 	return nil
```
Comment (rfranzke) [inline pkg/component/extensions/crds/crds.go]: same here

File: pkg/component/nodemanagement/machinecontrollermanager/crd.go [hunk 35]
```diff
@@ -82,3 +88,13 @@ func (c *crd) Destroy(ctx context.Context) error {
...
 }
+
+// Wait signals whether a CRD is ready or needs more time to be deployed.
+func (c *crd) Wait(ctx context.Context) error {
+	return kubernetesutils.WaitUntilCRDManifestsReady(ctx, c.client, maps.Keys(crdNameToManifest))
+}
+
+// WaitCleanup for destruction to finish and component to be fully removed. crdDeployer does not need to wait for cleanup.
+func (c *crd) WaitCleanup(_ context.Context) error {
+	return nil
```
Comment (rfranzke) [inline pkg/component/nodemanagement/machinecontrollermanager/crd.go]: same here

File: pkg/utils/kubernetes/crd_test.go [hunk 36]
```diff
@@ -0,0 +1,187 @@
...
+		It("should deploy a CRD", func() {
+			actualCRD := &apiextensionsv1.CustomResourceDefinition{}
+
+			err := crdDeployer.Deploy(ctx)
+			Expect(err).ToNot(HaveOccurred())
+
+			err = testClient.Get(ctx, client.ObjectKey{Name: readyCRD.Name}, actualCRD)
+			Expect(err).ToNot(HaveOccurred())
+			Expect(actualCRD.Name).To(Equal(readyCRD.Name))
+
```
Comment (rfranzke) [inline pkg/utils/kubernetes/crd_test.go]:
```suggestion
```

File: pkg/utils/kubernetes/crd_test.go [hunk 37]
```diff
@@ -0,0 +1,187 @@
...
+			actualCRD := &apiextensionsv1.CustomResourceDefinition{}
+
+			err := testClient.Create(ctx, readyCRD)
+			Expect(err).ToNot(HaveOccurred())
+
+			err = crdDeployer.Destroy(ctx)
+			Expect(err).ToNot(HaveOccurred())
+
+			Expect(testClient.Get(ctx, client.ObjectKey{Name: readyCRD.Name}, actualCRD)).To(matchers.BeNotFoundError())
+
```
Comment (rfranzke) [inline pkg/utils/kubernetes/crd_test.go]:
```suggestion
```

File: pkg/utils/kubernetes/crd_test.go [hunk 38]
```diff
@@ -0,0 +1,187 @@
...
+		})
+	})
+
+	Describe("#Destroy", func() {
+		It("should destroy a CRD", func() {
+			actualCRD := &apiextensionsv1.CustomResourceDefinition{}
+
+			err := testClient.Create(ctx, readyCRD)
+			Expect(err).ToNot(HaveOccurred())
```
Comment (rfranzke) [inline pkg/utils/kubernetes/crd_test.go]: Why not in one line? Expect(...).To(Succeed())?
Please take a look at other unit tests and the style there.
Apply this for similar occurrences in this PR.

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/utils/kubernetes/crd.go [hunk 39]
```diff
@@ -0,0 +1,138 @@
...
+	"github.com/gardener/gardener/pkg/utils/retry"
+)
+
+var (
+	// CRDWaitTimeout specifies the total time to wait for CRDs to become ready.
+	CRDWaitTimeout = 15 * time.Second
+)
+
+// CRDDeployer is a DeployWaiter that can deploy CRDs and wait for them to be ready.
+type CRDDeployer struct {
```
Comment (timuthy) [inline pkg/utils/kubernetes/crd.go]:
```suggestion
type crdDeployer struct {
```

File: pkg/utils/kubernetes/crd.go [hunk 40]
```diff
@@ -0,0 +1,138 @@
...
+	client            client.Client
+	applier           kubernetes.Applier
+	crdNameToManifest map[string]string
+}
+
+// NewCRDDeployer returns a DeployWaiter that can deploy CRDs and wait for them to be ready.
+func NewCRDDeployer(client client.Client, applier kubernetes.Applier, manifests []string) (*CRDDeployer, error) {
```
Comment (timuthy) [inline pkg/utils/kubernetes/crd.go]:
```suggestion
func NewCRDDeployer(client client.Client, applier kubernetes.Applier, manifests ...string) (component.DeployWaiter, error) {
```

File: pkg/component/autoscaling/vpa/crd.go [hunk 41]
```diff
@@ -71,3 +83,13 @@ func (v *vpaCRD) Destroy(ctx context.Context) error {
...
+
+// Wait signals whether a CRD is ready or needs more time to be deployed.
+func (v *vpaCRD) Wait(ctx context.Context) error {
+	return kubernetesutils.WaitUntilCRDManifestsReady(ctx, v.client, resourceNames)
+}
+
+// WaitCleanup for destruction to finish and component to be fully removed. crdDeployer does not need to wait for cleanup.
+func (v *vpaCRD) WaitCleanup(_ context.Context) error {
+	return nil
+}
```
Comment (timuthy) [inline pkg/component/autoscaling/vpa/crd.go]: It's not ideal that we don't use the CRDDeployer just because the `WaitCleanup` is different. You could for example work around by overloading `WaitCleanup` or by using [component.OpDestroyWithoutWait](https://github.com/gardener/gardener/blob/4d1d22130d77fa8f1391efd807ec205964482a73/pkg/component/deploy.go#L39).

File: pkg/component/etcd/etcd/crd.go (hunk 33 above)
Comment (timuthy) [reply pkg/component/etcd/etcd/crd.go]: Wouldn't it even better to use the `CRDDeployer`?

File: pkg/utils/kubernetes/crd_test.go (hunk 18 above)
Comment (timuthy) [reply pkg/utils/kubernetes/crd_test.go]: If you say any kind of error, then you don't needed to match any string at all. If the string is too bulky then match a more precise string?

**Review by timuthy — CHANGES_REQUESTED** [review]

File: pkg/utils/kubernetes/crd.go (hunk 40 above)
Comment (tobschli) [reply pkg/utils/kubernetes/crd.go]: Thank you for the oversight :)
Because of the circular dependencies we talked about, I will move this to `components/crddeployer`

Although, I am not too sure about the change to a variadic function, as most of the deployers have the manifests in slices anyways

File: pkg/component/autoscaling/vpa/crd.go (hunk 41 above)
Comment (tobschli) [reply pkg/component/autoscaling/vpa/crd.go]: Like we talked about, I will use the generic `CRDDeployer` as an anonymous field for the individual deployers.
An exception is the `istio` deployer, where I think, it does not make much sense to use the generic deployer, as the `istio` deployer renders a Helm chart. To render it, it does not use the `Applier` that the other deployers use, but a `ChartApplier`.
I think it would be too big of a redundancy to possibly create one and pass it to the `istio` deployer, solely for the reason of `Wait` and `WaitCleanup`.
What do you think?

**Comment by tobschli** [comment]
Amongst other issues with this PR at the current state, I observed that cleaning up CRDs might actually be something we _don't_ want, as e.g. the e2e test suite tries to collect `Etcd`s to export into the artifacts.

With the proposed changed of completely cleaning up CRDs, this turns out to be problematic ([ref](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/10639/pull-gardener-e2e-kind-upgrade/1851584232573374464#1:build-log.txt%3A1913))

**Comment by tobschli** [comment]
Due to the increasing complexity of the deflake, this PR will be closed in favor of another PR that only implements the generic `CRDDeployer`, which will only be used for the `prometheusoperator` CRD deployer.

Once this is done, the generic `CRDDeployer' should be adapted to the other deployers to achieve consistency.
